File: py/data/as_commands.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict

from data import as_commands_pb2
from google.protobuf.json_format import MessageToJson, Parse


@dataclass
class ASCommands:
    steerAngleDegrees: float = 0.0
    pedalThrottle: float = 0.0
    pedalBrakes: float = 0.0
    
    _proto_message: as_commands_pb2.ASCommands = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = as_commands_pb2.ASCommands()

    def _populate_proto(self):
        self._proto_message.steerAngleDegrees = self.steerAngleDegrees
        self._proto_message.pedalThrottle = self.pedalThrottle
        self._proto_message.pedalBrakes = self.pedalBrakes

    @classmethod
    def _from_proto(cls, proto_message) -> "ASCommands":
        return cls(
            steerAngleDegrees = proto_message.steerAngleDegrees,
            pedalThrottle = proto_message.pedalThrottle,
            pedalBrakes = proto_message.pedalBrakes,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "ASCommands":
        message = as_commands_pb2.ASCommands()
        message.ParseFromString(data)
        return cls(
            steerAngleDegrees = message.steerAngleDegrees,
            pedalThrottle = message.pedalThrottle,
            pedalBrakes = message.pedalBrakes,
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "ASCommands":
        message = as_commands_pb2.ASCommands()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())
class Status(Enum):
    status_disabled = 0
    status_enabled = 1


@dataclass
class ASStatus:
    steerStatus: Status = None
    throttleStatus: Status = None
    brakesStatus: Status = None
    
    _proto_message: as_commands_pb2.ASStatus = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = as_commands_pb2.ASStatus()

    def _populate_proto(self):
        self._proto_message.steerStatus = self.steerStatus.value
        self._proto_message.throttleStatus = self.throttleStatus.value
        self._proto_message.brakesStatus = self.brakesStatus.value

    @classmethod
    def _from_proto(cls, proto_message) -> "ASStatus":
        return cls(
            steerStatus = Status._from_proto(proto_message.steerStatus),
            throttleStatus = Status._from_proto(proto_message.throttleStatus),
            brakesStatus = Status._from_proto(proto_message.brakesStatus),
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "ASStatus":
        message = as_commands_pb2.ASStatus()
        message.ParseFromString(data)
        return cls(
            steerStatus = Status(message.steerStatus),
            throttleStatus = Status(message.throttleStatus),
            brakesStatus = Status(message.brakesStatus),
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "ASStatus":
        message = as_commands_pb2.ASStatus()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

File: src/data/as_commands.cpp
#include "data/as_commands.h"

#include <google/protobuf/util/json_util.h>
#include <tuple>

namespace Serializers
{
namespace Data
{
ASCommands::ASCommands(const PbData::ASCommands& protobuf) {
    steerAngleDegrees = protobuf.steerangledegrees();
    pedalThrottle = protobuf.pedalthrottle();
    pedalBrakes = protobuf.pedalbrakes();
}

ASCommands::operator PbData::ASCommands() const {
    PbData::ASCommands ret;
    ret.set_steerangledegrees(steerAngleDegrees);
    ret.set_pedalthrottle(pedalThrottle);
    ret.set_pedalbrakes(pedalBrakes);
    return ret;
}

std::string ASCommands::serializeAsJsonString() const {
    PbData::ASCommands protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string ASCommands::serializeAsProtobufString() const {
    PbData::ASCommands protobuf(*this);
    return protobuf.SerializeAsString();
}

bool ASCommands::deserializeFromJsonString(const std::string& str) {
    PbData::ASCommands protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool ASCommands::deserializeFromProtobufString(const std::string& str) {
    PbData::ASCommands protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

ASStatus::ASStatus(const PbData::ASStatus& protobuf) {
    steerStatus = static_cast<Status>(protobuf.steerstatus());
    throttleStatus = static_cast<Status>(protobuf.throttlestatus());
    brakesStatus = static_cast<Status>(protobuf.brakesstatus());
}

ASStatus::operator PbData::ASStatus() const {
    PbData::ASStatus ret;
    ret.set_steerstatus(static_cast<PbData::Status>(steerStatus));
    ret.set_throttlestatus(static_cast<PbData::Status>(throttleStatus));
    ret.set_brakesstatus(static_cast<PbData::Status>(brakesStatus));
    return ret;
}

std::string ASStatus::serializeAsJsonString() const {
    PbData::ASStatus protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string ASStatus::serializeAsProtobufString() const {
    PbData::ASStatus protobuf(*this);
    return protobuf.SerializeAsString();
}

bool ASStatus::deserializeFromJsonString(const std::string& str) {
    PbData::ASStatus protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool ASStatus::deserializeFromProtobufString(const std::string& str) {
    PbData::ASStatus protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}
}
}